nixos/modules/system/boot/luksroot.nix

```diff
@@ -207,19 +207,47 @@ let
 
     # LUKS
     open_normally() {
+        keyFileFailed=true
         ${if (dev.keyFile != null) then ''
         if wait_target "key file" ${dev.keyFile}; then
             ${csopen} --key-file=${dev.keyFile} \
               ${optionalString (dev.keyFileSize != null) "--keyfile-size=${toString dev.keyFileSize}"} \
               ${optionalString (dev.keyFileOffset != null) "--keyfile-offset=${toString dev.keyFileOffset}"}
+            cs_status=$?
+            if [ $cs_status -ne 0 ]; then
+               echo "Key File ${dev.keyFile} failed!"
+            else
+               keyFileFailed=false
+            fi
         else
-            ${if dev.fallbackToPassword then "echo" else "die"} "${dev.keyFile} is unavailable"
-            echo " - failing back to interactive password prompt"
-            do_open_passphrase
+            echo "Key file not found!"
         fi
         '' else ''
-        do_open_passphrase
+           echo "No key file given"
         ''}
+
+        if [ "$keyFileFailed" = true ]; then
+            ${if (dev.tryEmptyPassphrase) then ''
+                 echo "Trying empty passphrase!"
+                 echo "" | ${csopen}
+                 cs_status=$?
+                 if [ $cs_status -eq 0 ]; then
+                    echo "Empty passphrase succeeded"
+                 else
+                    ${if dev.fallbackToPassword then ''
+                      echo "${dev.keyFile} is unavailable and empty passphrase failed"
+                      echo " - falling back to interactive password prompt"
+                      do_open_passphrase
+                    '' else "die"}
+                 fi
+            '' else ''
+                 ${if dev.fallbackToPassword then ''
+                      echo "${dev.keyFile} is unavailable"
+                      echo " - falling back to interactive password prompt"
+                      do_open_passphrase
+                 '' else "die"}
+            ''}
+        fi
```

Comment:
I'm not entirely sure this block correctly handles the case where the user just wants to enter a password? `dev.fallbackToPassword` defaults to `false`, and in the `else` cases for those checks, you have `die` in each of them. Have you tested this PR with regular password entry?

Also, seems like a lot more echo'ing will happen basically no matter what, including messages about key files even when none is configured.

I haven't tested these questions myself, so let me know if I'm just following the code flow wrong.

Reply:
Looks like you are correct. I think changing `dev.fallbackToPassword` to default to `true` is the simplest fix and is the action most users will want. If a user doesn't want this then they can explicitly set this option to false

Reply:
@RaitoBezarius @ElvishJerricco @peterwilli Do you guys agree with this as the default behavior?

Reply:
I do not agree with changing the default behavior. Neither possible default is inherently better than the other, so it should stay the same.

Reply:
I'm pretty sure what I described is what already happens: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/luksroot.nix#L221

Reply:
@therishidesai That is in the `else` block. That only happens if `keyFile == null`. i.e. when `keyFile == null`, always use passphrase, but when `keyFile != null`, only fallback to passphrase if `fallbackToPassword == true`. This is the behavior we should maintain, and we should not change the default values of these things.

Reply:
@therishidesai actually I feel like this entire code block is overengineered. I think this would have done the trick much more easily (untested)

```patch
diff --git a/nixos/modules/system/boot/luksroot.nix b/nixos/modules/system/boot/luksroot.nix
index bbdb26f975e..9ca26b44536 100644
--- a/nixos/modules/system/boot/luksroot.nix
+++ b/nixos/modules/system/boot/luksroot.nix
@@ -158,6 +158,13 @@ let
       wait_target "header" ${dev.header} || die "${dev.header} is unavailable"
     ''}
 
+    try_empty_passphrase() {
+        ${if dev.tryEmptyPassphrase then ''
+            echo " - trying empty passphrase..."
+            ${csopen} --key-file=/dev/null
+        '' else "return 1"}
+    }
+
     do_open_passphrase() {
         local passphrase
 
@@ -213,12 +220,16 @@ let
               ${optionalString (dev.keyFileSize != null) "--keyfile-size=${toString dev.keyFileSize}"} \
               ${optionalString (dev.keyFileOffset != null) "--keyfile-offset=${toString dev.keyFileOffset}"}
         else
-            ${if dev.fallbackToPassword then "echo" else "die"} "${dev.keyFile} is unavailable"
-            echo " - failing back to interactive password prompt"
-            do_open_passphrase
+            if ! try_empty_passphrase; then
+                ${if dev.fallbackToPassword then "echo" else "die"} "${dev.keyFile} is unavailable"
+                echo " - failing back to interactive password prompt"
+                do_open_passphrase
+            fi
         fi
         '' else ''
-        do_open_passphrase
+        if ! try_empty_passphrase; then
+            do_open_passphrase
+        fi
         ''}
     }
 

```